### PR TITLE
[AutoParallel] adapt for 2d laplace

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

New features
### PR changes

Others
### Describe

- `backward.py`: add default distributed context of auto_parallel to record the mapping between grad var name and var name.
- `completion.py`: update forward and backward phase with `dist_op_context.grad_var_to_var`.
- `partitioner.py`: update init mapping to record all the ops before loss op.
- `dist_default.py`: 
    * compatible func:  considering `XShape` dims_mapping when `XShape` in op.desc.inputs.
    * need_gradient_allreduce: get allreduce var from `dist_op_context.grad_var_to_var`.
- `utils.py`: update `set_grad_var_shape` func to reset grad var shape when grad time is bigger than 1.
- `engine.py`: `exe.run(start_program)` only once for every variable.

### CODE REVIEW

**Review by aoyulong — DISMISSED** [review]
This pr is just an ad-hoc solution for now and will be replaced by a new implementation. But we will not change the engine api.

File: python/paddle/fluid/backward.py [hunk 1]
```diff
@@ -1081,6 +1099,16 @@ def _append_backward_ops_(block,
         rename_var_map(dict): used to associate target_grad var name with first grad_op input name.
             Only used in for high order gradient.
     """
+
+    # Build the mapping between the forward op and backward op (Only for auto parallel)
+    def update_distop_context(distop_context, op_grad_to_var,
```
Comment (JiabinYang) [inline python/paddle/fluid/backward.py:1104]: maybe name this as `update_dist_op_context`

File: python/paddle/fluid/backward.py [hunk 2]
```diff
@@ -1118,11 +1146,18 @@ def _append_backward_ops_(block,
         # Getting op's corresponding grad_op
         grad_op_desc, op_grad_to_var = core.get_grad_op_desc(
             op.desc, cpt.to_text(no_grad_dict[block.idx]), grad_sub_block_list)
+
         # Build the mapping between the forward op and backward op (Only for auto parallel)
         if distop_context is not None:
```
Comment (JiabinYang) [inline python/paddle/fluid/backward.py:1151]: same

File: python/paddle/fluid/backward.py [hunk 3]
```diff
@@ -1187,8 +1227,14 @@ def _append_backward_ops_(block,
             grad_op_descs.extend(grad_op_desc)
             grad_to_var.update(op_grad_to_var)
 
+    # record mapping bewteen grad var name and var name (Only for auto parallel)
+    grad_var_to_var = None
+    if distop_context is not None:
+        grad_var_to_var = distop_context.grad_var_to_var[
```
Comment (JiabinYang) [inline python/paddle/fluid/backward.py:1233]: rename to `dist_op_context.grad_var_to_var_maps` ?

**Review by JiabinYang — APPROVED** [review]
some typo issue, LGTM for backward.py

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM for `set_tests_properties(test_high_order_grad PROPERTIES LABELS "RUN_TYPE=EXCLUSIVE" TIMEOUT 50)`

File: python/paddle/fluid/backward.py (hunk 3 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/backward.py:1233]: it will be updated in the next pr.

File: python/paddle/fluid/backward.py (hunk 2 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/backward.py:1151]: it will be updated in the next pr.

File: python/paddle/fluid/backward.py (hunk 1 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/backward.py:1104]: it will be updated in the next pr.

**Pull request merged**